spices/SPICE-0001-import-graph-analyzer-api.adoc

```diff
@@ -0,0 +1,336 @@
+
+        Map<URI, URI> getResolvedUris();
+    }
+}
+
+----
+
+=== Example
+
+This Pkl file:
+
+./path/to/my/foo.pkl
+[source,pkl]
+----
+import "@bar/Bar.pkl" // <1>
+import "package://example.com/bar@1.0.0#/Bar.pkl" // <2>
+import "pkl:reflect" // <3>
+import "..." // <4>
+import* "subdir/*.pkl" <5>
+
+res = if (foo) import("foo1.pkl") else import("foo2.pkl") // <6>
+----
+<1> local dependency import of package://example.com/bar@1.0.0
+<2> direct package import
+<3> stdlib import
+<4> triple-dot import (resolves to /path/to/foo.pkl)
+<5> globbed import (resolves to /path/to/my/subdir/foo.pkl)
+<6> conditional import; both are added to the dependency graph
+
+Produces the following result:
+
+[source,pkl]
+----
+imports {
+  ["projectpackage://example.com/bar@1.0.0#/Bar.pkl"] {} // <1>
+  ["file:///foo/bar/Bar.pkl"] {}
+  ["package://example.com/bar@1.0.0#/Bar.pkl"] {}
+  ["pkl:reflect"] {
+    "pkl:base"
+  }
+  ["pkl:base"] {
+    "pkl:jsonnet"
+    "pkl:xml"
+    "pkl:protobuf"
+  }
+  ["pkl:jsonnet"] {}
+  ["pkl:xml"] {}
+  ["pkl:protobuf"] {
+    "pkl:reflect"
+  }
+  ["file:///path/to/foo.pkl"] {}
+  ["file:///path/to/my/foo1.pkl"] {}
+  ["file:///path/to/my/foo2.pkl"] {}
+  ["file:///path/to/my/subdir/foo.pkl"] {}
+}
+resolvedUris {
+  ["projectpackage://example.com/bar@1.0.0#/Bar.pkl"] = "file:///path/to/local/bar/Bar.pkl" // <2>
+  ["file:///foo/bar/Bar.pkl"] = "file:///foo/bar/bar.pkl" // <3>
+  ["package://example.com/bar@1.0.0#/Bar.pkl"] = "package://example.com/bar@1.0.0#/Bar.pkl" // <4>
+  ["pkl:reflect"] = "pkl:reflect"
+  ["pkl:jsonnet"] = "pkl:jsonnet"
+  ["pkl:xml"] = "pkl:xml"
+  ["pkl:protobuf"] = "pkl:protobuf"
+  ["file:///path/to/foo.pkl"] = "file:///path/to/foo.pkl"
+  ["file:///path/to/my/foo1.pkl"] = "file:///path/to/my/foo1.pkl"
+  ["file:///path/to/my/foo2.pkl"] = "file:///path/to/my/foo2.pkl"
+  ["file:///path/to/my/subdir/foo.pkl"] = "file:///path/to/my/subdir/foo.pkl"
+}
+----
+<1> dependency notation turns into `projectpackage://`.
+<2> local dependency's resolve URI is a local file.
+<3> module URIs that don't resolve to anything further stay the same.
+<4> package URIs stay the same.
+
+=== Gradle Plugin Changes
+
+The input files for various Pkl Gradle tasks will switch to using the Java `ImportAnalyzer` to build its input files.
+
+The set of input files will be `resolvedUris` whose keys start with `\file:///`.
+All other modules are assumed to be external to the machine and cache-able.
+
+The `transitiveModules` property is removed.
+
+== Future directions
+
+=== `pkl:anaylze` standard library module
+
+This is a starting point for the `pkl.analyze` stdlib module.
+Other forms of static analysis can be added to it as the need grows.
+
+=== Support in other language bindings
+
+This feature introduces `PklAnalyzer`, a new Java API that can analyze the imports of modules without requiring the evaluator.
+This is a capability that isn't being added to other language bindings.
```

Comment:
In my opinion, the longer-term goal should be to make every feature available to language bindings and to have no privileged Java APIs.

Reply:
Definitely agreed! But, I am setting that as out of scope for this SPICE; this can be a future improvement.